src/app/interaction-model/Model.h

```diff
@@ -0,0 +1,117 @@
+
+    /// List reading has specific handling logic:
+    ///   `state` contains in/out data about the current list reading. MUST start with kInvalidListIndex on first call
+    ///
+    /// Return codes:
+    ///   CHIP_ERROR_MORE_LIST_DATA_AVAILABLE (NOTE: new error defined for this purpose)
+    ///      - partial data written to the destination
+    ///      - destination will contain AT LEAST one valid list entry fully serialized
+    ///      - destination will be fully valid (it will be rolled back on partial list writes)
+    ///   CHIP_IM_GLOBAL_STATUS(code):
+    ///      - error codes that are translatable in IM status codes (otherwise we expect Failure to be reported)
+    ///      - In particular, some handlers rely on special handling for:
+    ///        - `UnsupportedAccess`  - for ACL checks (e.g. wildcard expansion may choose to skip these)
+    ///      - to check for this, CHIP_ERROR provides:
+    ///        - ::IsPart(ChipError::SdkPart::kIMGlobalStatus) -> bool
+    ///        - ::GetSdkCode() -> uint8_t to translate to the actual code
+    virtual CHIP_ERROR ReadAttribute(const ReadAttributeRequest & request, ReadState & state,
+                                     chip::TLV::TLVWriter & attribute_data) = 0;
+
+    /// Requests a write of an attribute.
+    ///
+    /// When this is invoked, caller is expected to have already done some validations:
+    ///    - cluster `data version` has been checked for the incoming request if applicable
+    ///
+    /// List operation support:
+    ///    - the first list write will have `request.writeFlags.Has(WriteFlags::kListBegin)`
+    ///    - the last list write will have `request.writeFlags.Has(WriteFlags::kListEnd)`
+    ///    - the last list write MAY have empty data (no list items)
+    ///
+    /// When `request.writeFlags.Has(WriteFlags::kForceInternal)` the request is from an internal app update
+    /// and SHOULD bypass some internal checks (like timed enforcement, potentially read-only restrictions)
+    ///
+    /// Return codes
+    ///   CHIP_IM_GLOBAL_STATUS(code):
+    ///       - error codes that are translatable to specific IM codes
+    ///       - in particular, the following codes are interesting/expected
+    ///         - `UnsupportedWrite` for attempts to write read-only data
+    ///         - `UnsupportedAccess` for ACL failures
+    ///         - `NeedsTimedInteraction` for writes that are not timed however are required to be so
+    virtual CHIP_ERROR WriteAttribute(const WriteAttributeRequest & request, chip::TLV::TLVReader & attribute_data) = 0;
+
+    /// `responder` is used to send back the reply.
+    ///    - calling Reply() or ReplyAsync() will let the application control the reply
+    ///    - returning a CHIP_NO_ERROR without reply/reply_async implies a Status::Success reply without data
+    ///    - returning a CHIP_*_ERROR implies an error reply (error and data are mutually exclusive)
+    ///
```

Comment:
Added some documentation, hopefully in a better manner. In most cases it seems to me clients do not care sufficiently if a reply was fully sent to the other side since there is nothing they can actually do except logging, which the code would already do anyway.